samples/extensions/khr/externalmemory/main.cpp

```diff
@@ -0,0 +1,559 @@
+
+        VkMemoryAllocateInfo memory_alloc_info_x{};
+        memory_alloc_info_x.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+        memory_alloc_info_x.pNext = &export_memory_alloc_info;
+        memory_alloc_info_x.allocationSize = mem_requirements_x.size;
+        memory_alloc_info_x.memoryTypeIndex = find_vk_memory_type(
+            vk_physical_device, mem_requirements_x.memoryTypeBits,
+            VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT
+                | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
+
+        VkMemoryAllocateInfo memory_alloc_info_y{};
+        memory_alloc_info_y.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+        memory_alloc_info_y.pNext = &export_memory_alloc_info;
+        memory_alloc_info_y.allocationSize = mem_requirements_y.size;
+        memory_alloc_info_y.memoryTypeIndex = find_vk_memory_type(
+            vk_physical_device, mem_requirements_y.memoryTypeBits,
+            VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT
+                | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
+
+        // Allocate and bind memory.
+        VkDeviceMemory vk_buf_x_memory, vk_buf_y_memory;
+        VK_CHECK(vkAllocateMemory(vk_device, &memory_alloc_info_x, nullptr,
+                                  &vk_buf_x_memory));
+        VK_CHECK(vkAllocateMemory(vk_device, &memory_alloc_info_y, nullptr,
+                                  &vk_buf_y_memory));
+
+        VK_CHECK(vkBindBufferMemory(vk_device, vk_buf_x, vk_buf_x_memory, 0));
+        VK_CHECK(vkBindBufferMemory(vk_device, vk_buf_y, vk_buf_y_memory, 0));
+
+        // Map memory.
+        void *vk_arr_x, *vk_arr_y;
+        VK_CHECK(vkMapMemory(vk_device, vk_buf_x_memory, 0, VK_WHOLE_SIZE, 0,
+                             &vk_arr_x));
+        VK_CHECK(vkMapMemory(vk_device, vk_buf_y_memory, 0, VK_WHOLE_SIZE, 0,
+                             &vk_arr_y));
+
+        memcpy(vk_arr_x, arr_x.data(), sizeof(cl_float) * length);
+        memcpy(vk_arr_y, arr_y.data(), sizeof(cl_float) * length);
+
+        // Get Vulkan external memory file descriptors for accessing external
+        // memory with OpenCL.
+        VkMemoryGetFdInfoKHR fd_info_x{};
+        fd_info_x.sType = VK_STRUCTURE_TYPE_MEMORY_GET_FD_INFO_KHR;
+        fd_info_x.pNext = nullptr;
+        fd_info_x.memory = vk_buf_x_memory;
+        fd_info_x.handleType = vk_external_memory_handle_type;
+        int fd_x;
+
+        VkMemoryGetFdInfoKHR fd_info_y{};
+        fd_info_y.sType = VK_STRUCTURE_TYPE_MEMORY_GET_FD_INFO_KHR;
+        fd_info_y.pNext = nullptr;
+        fd_info_y.memory = vk_buf_y_memory;
+        fd_info_y.handleType = vk_external_memory_handle_type;
+        int fd_y;
+
+        // We need to get the pointer to the vkGetMemoryFdKHR function because
+        // it's from extension VK_KHR_external_memory_fd.
+        PFN_vkGetMemoryFdKHR vkGetMemoryFdKHR =
+            (PFN_vkGetMemoryFdKHR)vkGetDeviceProcAddr(vk_device,
+                                                      "vkGetMemoryFdKHR");
```

Comment:
This is the function that is needed for Linux and file descriptors, but a different function is needed for Windows and NT handles.  Has this sample been tested on Windows?

Reply:
Yes, this sample has been tested on Windows, but we got 
```
Test #31: externalmemory_Release ...........Exit code 0xc0000135
***Exception:   0.01 sec

Test #32: externalmemorycpp_Release ........Exit code 0xc0000135
***Exception:   0.01 sec
```
In the latest commit, I have added the logic to get the function pointer and call the correct function on Windows, but the runtime fail is still there: https://github.com/StreamHPC/OpenCL-SDK/actions/runs/7216235279/job/19662376992. This runtime error also appears on MacOS, although for now both tests are disabled on CI ([macos' here](https://github.com/StreamHPC/OpenCL-SDK/blob/external-mem-sample/.github/workflows/presubmit.yml#L743-L745) and [windows' here](https://github.com/StreamHPC/OpenCL-SDK/blob/external-mem-sample/.github/workflows/presubmit.yml#L571-L578)).

Given that it works fine in Linux, we assumed this failure could be due to no appropriate Vulkan device present on the failing test machines from CI. Perhaps you have more insights about this?

Reply:
I think this is still failing because the handle type needs to be updated for Windows:

```c++
            (cl_mem_properties)CL_EXTERNAL_MEMORY_HANDLE_OPAQUE_FD_KHR,
#ifdef _WIN32
            (cl_mem_properties)handle_y,
#else
            (cl_mem_properties)fd_y,
#endif
```

Reply:
I think this still needs to be fixed.  Have these samples been tested on Windows?

Reply:
Forgot to answer this, sorry. The handle type is allegedly fixed, but the example still fails on Windows. Right now is disabled on CI, but I tested it on my personal fork https://github.com/Beanavil/OpenCL-SDK/actions/runs/9563524333/job/26362215011.